src/main/java/com/google/jenkins/plugins/computeengine/PreemptedCheckCallable.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.jenkins.plugins.computeengine;
+
+import com.google.api.client.http.GenericUrl;
+import com.google.api.client.http.HttpHeaders;
+import com.google.api.client.http.HttpRequest;
+import com.google.api.client.http.HttpResponse;
+import com.google.api.client.http.HttpTransport;
+import com.google.api.client.http.javanet.NetHttpTransport;
+import hudson.model.TaskListener;
+import java.io.IOException;
+import jenkins.security.MasterToSlaveCallable;
+import org.apache.commons.io.Charsets;
+import org.apache.commons.io.IOUtils;
+
+/**
+ * Slave callback class checking if instance was preempted. All of code here is serialized and
+ * executed on node.
+ */
+final class PreemptedCheckCallable extends MasterToSlaveCallable<Boolean, IOException> {
+  private static final String METADATA_SERVER_URL =
+      "http://metadata.google.internal/computeMetadata/v1/instance/preempted?wait_for_change=true";
+
+  private final TaskListener listener;
+
+  /**
+   * Callback constructor.
+   *
+   * @param listener Node listener on which we can add extra information from slave side.
+   */
+  PreemptedCheckCallable(TaskListener listener) {
+    this.listener = listener;
+  }
+
+  /**
+   * Actual callback code, executed on node side. Checks in Google metadata server if instance was
+   * preempted.
+   *
+   * <p>See
+   * https://cloud.google.com/compute/docs/instances/create-start-preemptible-instance#detecting_if_an_instance_was_preempted
+   *
+   * @return True if node was preempted.
+   * @throws IOException Exception when calling Google metadata API
+   */
+  @Override
+  public Boolean call() throws IOException {
+    HttpRequest request = createMetadataRequest();
+    listener.getLogger().println("Preemptive instance, listening to metadata for preemption event");
+    HttpResponse response = request.execute();
+    final String result = IOUtils.toString(response.getContent(), Charsets.UTF_8);
+    listener.getLogger().println("Got preemption event " + result);
+    response.disconnect();
+    return "TRUE".equals(result);
+  }
+
+  private HttpRequest createMetadataRequest() throws IOException {
+    HttpTransport transport = new NetHttpTransport();
```

Comment:
Taking another look at this. This won't work as is. This logic needs to be consolidated with the ClientFactory, and the request needs to be built using the GoogleClientRequestInitializer similar to the approached used for the ComputeClient. A number of our customers are running their masters on-prem and thus won't have the GCP SA baked into the VM metadata, which is what would be required for this to work as is.

Reply:
This is what is causing the ITs to fail for me btw, as I'm running them on a VM not in GCE.

Reply:
But this code is running on `slave` side.. not master.. so  it should work..
Could you explain how would you like it to run?

Anything except calling metadata will not work cause all other sources have delay..

Reply:
And I run tests from my own computer.. not in GCE for sure and it works..
But maybe there is something wrong here..
Cause it seems like its not catching what it should..

Reply:
@craigdbarber Did you found why tests on your side are failing?

Reply:
Not yet. Both @stephenashank and I are encountering the same problem. Could you do us a favor an run this command: gcloud projects get-iam-policy <YOUR GCLOUD PROJECT>  \
--flatten="bindings[].members" \
--format='table(bindings.role)' \
--filter="bindings.members:<YOUR SERVICE ACCOUNT>"
And paste the results into this thread. Perhaps that will help us get to the bottom of this.

Reply:
```
ROLE
roles/editor
```

Reply:
any luck? can try to get this going on my own environment